File: ai-manus/backend/app/interfaces/schemas/claw.py
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime
from app.domain.models.claw import ClawStatus, ClawMessage, ClawAttachment


class ClawResponse(BaseModel):
    """Claw instance response schema"""
    id: str
    user_id: str
    status: ClawStatus
    container_name: Optional[str] = None
    error_message: Optional[str] = None
    expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    @staticmethod
    def from_claw(claw) -> 'ClawResponse':
        return ClawResponse(
            id=claw.id,
            user_id=claw.user_id,
            status=claw.status,
            container_name=claw.container_name,
            error_message=claw.error_message,
            expires_at=claw.expires_at,
            created_at=claw.created_at,
            updated_at=claw.updated_at,
        )


class ClawApiKeyResponse(BaseModel):
    """API key response schema"""
    api_key: str


class ClawChatRequest(BaseModel):
    """Chat request schema"""
    message: str
    session_id: str = "default"


class ClawChatChunk(BaseModel):
    """Chat chunk event schema"""
    type: str
    content: Optional[str] = None
    stop_reason: Optional[str] = None
    error: Optional[str] = None


class ClawAttachmentSchema(BaseModel):
    """File attachment in a chat message"""
    file_id: str
    filename: str
    content_type: Optional[str] = None
    size: int = 0
    file_url: Optional[str] = None

    @staticmethod
    def from_domain(att: ClawAttachment) -> 'ClawAttachmentSchema':
        return ClawAttachmentSchema(
            file_id=att.file_id,
            filename=att.filename,
            content_type=att.content_type,
            size=att.size,
            file_url=att.file_url,
        )


class ClawMessageSchema(BaseModel):
    """A single chat message"""
    role: str
    content: str = ""
    timestamp: int
    attachments: Optional[List[ClawAttachmentSchema]] = None

    @staticmethod
    def from_domain(msg: ClawMessage) -> 'ClawMessageSchema':
        return ClawMessageSchema(
            role=msg.role,
            content=msg.content,
            timestamp=msg.timestamp,
            attachments=[ClawAttachmentSchema.from_domain(a) for a in msg.attachments] if msg.attachments else None,
        )


class ClawHistoryResponse(BaseModel):
    """Chat history response"""
    messages: List[ClawMessageSchema]

File: ai-manus/frontend/src/api/claw.ts
import { apiClient, ApiResponse, BASE_URL } from './client';
import { getStoredToken } from './auth';

export type ClawStatus = 'creating' | 'running' | 'stopped' | 'error';

export interface Claw {
  id: string;
  user_id: string;
  status: ClawStatus;
  container_name?: string;
  error_message?: string;
  expires_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClawApiKey {
  api_key: string;
}

export interface ClawEvent {
  type: 'text' | 'done' | 'error' | 'file' | 'catchup' | 'heartbeat' | 'status';
  content?: string;
  stop_reason?: string;
  error?: string;
  status?: ClawStatus;
  file_id?: string;
  filename?: string;
  content_type?: string;
  size?: number;
  upload_date?: string;
  file_url?: string;
}

export interface ClawChatAttachment {
  file_id: string;
  filename: string;
  content_type?: string;
  size: number;
  file_url?: string;
}

export interface ClawChatMessage {
  role: 'user' | 'assistant' | 'attachments';
  content: string;
  timestamp: number;
  attachments?: ClawChatAttachment[];
}

// ---- REST endpoints ----

export async function getClaw(): Promise<Claw> {
  const response = await apiClient.get<ApiResponse<Claw>>('/claw');
  return response.data.data;
}

export async function createClaw(): Promise<Claw> {
  const response = await apiClient.post<ApiResponse<Claw>>('/claw');
  return response.data.data;
}

export async function deleteClaw(): Promise<void> {
  await apiClient.delete<ApiResponse<{}>>('/claw');
}

export async function getClawApiKey(): Promise<string> {
  const response = await apiClient.get<ApiResponse<ClawApiKey>>('/claw/api-key');
  return response.data.data.api_key;
}

export async function getClawHistory(): Promise<ClawChatMessage[]> {
  const response = await apiClient.get<ApiResponse<{ messages: ClawChatMessage[] }>>('/claw/history');
  return response.data.data.messages;
}

// ---- WebSocket connection ----

export interface ClawWSCallbacks {
  onEvent: (event: ClawEvent) => void;
  onOpen?: () => void;
  onClose?: () => void;
}

/**
 * Manages a persistent WebSocket connection to the Claw backend.
 * Auto-reconnects on disconnect with exponential backoff.
 */
export class ClawWebSocket {
  private ws: WebSocket | null = null;
  private callbacks: ClawWSCallbacks;
  private closed = false;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private reconnectDelay = 1000;

  constructor(callbacks: ClawWSCallbacks) {
    this.callbacks = callbacks;
    this.connect();
  }

  private connect() {
    if (this.closed) return;

    const wsBase = BASE_URL.replace(/^http/, 'ws');
    const token = getStoredToken();
    const params = token ? `?token=${encodeURIComponent(token)}` : '';
    const url = `${wsBase}/claw/ws${params}`;

    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      this.reconnectDelay = 1000;
      this.callbacks.onOpen?.();
    };

    this.ws.onmessage = (e) => {
      try {
        const data: ClawEvent = JSON.parse(e.data);
        if (data.type !== 'heartbeat') {
          this.callbacks.onEvent(data);
        }
      } catch {
        // ignore
      }
    };

    this.ws.onclose = () => {
      this.callbacks.onClose?.();
      this.scheduleReconnect();
    };

    this.ws.onerror = () => {
      this.ws?.close();
    };
  }

  private scheduleReconnect() {
    if (this.closed) return;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectDelay = Math.min(this.reconnectDelay * 2, 30000);
      this.connect();
    }, this.reconnectDelay);
  }

  /**
   * Send a chat message through the WebSocket, optionally with file attachments.
   */
  send(message: string, sessionId: string = 'default', fileIds?: string[]) {
    if (this.ws?.readyState === WebSocket.OPEN) {
      const payload: Record<string, unknown> = { type: 'chat', message, session_id: sessionId };
      if (fileIds && fileIds.length > 0) {
        payload.file_ids = fileIds;
      }
      this.ws.send(JSON.stringify(payload));
    }
  }

  /**
   * Close the connection permanently (no auto-reconnect).
   */
  disconnect() {
    this.closed = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.ws?.close();
    this.ws = null;
  }

  get isConnected() {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}
